src/main/java/com/epam/eco/commons/kafka/helpers/BiDirectionalTopicRecordFetcher.java

```diff
@@ -0,0 +1,366 @@
+        validatePartitionTimestamps(partitionTimestamps);
+        Validate.isTrue(limit > 0, "Limit is invalid");
+        Validate.isTrue(timeoutInMs > 0, "Timeout is invalid");
+
+        try (KafkaConsumer<K, V> consumer = new KafkaConsumer<>(consumerConfig)) {
+            return doFetchByTimestamps(consumer, partitionTimestamps, limit, filter, timeoutInMs, direction);
+        }
+    }
+
+
+    private RecordFetchResult<K, V> doFetchByTimestamps(KafkaConsumer<K, V> consumer, Map<TopicPartition, Long> partitionTimestamps, long limit, Predicate<ConsumerRecord<K, V>> filter, long timeoutMs, FetchDirection direction) {
+        Map<TopicPartition, OffsetAndTimestamp> offsetsForTimes = consumer.offsetsForTimes(partitionTimestamps);
+
+        Map<TopicPartition, Long> offsets = offsetsForTimes.entrySet().stream().filter(entry -> entry.getValue() != null).collect(Collectors.toMap(Map.Entry::getKey, entry -> entry.getValue().offset()));
+
+        return direction == FetchDirection.FORWARD  ? doFetchByOffsets(consumer, offsets, limit, filter, timeoutMs) : doReverseFetchByOffsets(consumer, offsets, limit, filter, timeoutMs);
+    }
+
+
+
+    protected RecordFetchResult<K, V> doReverseFetchByOffsets(KafkaConsumer<K, V> consumer,
+                                                            Map<TopicPartition, Long> offsets,
+                                                            long limit,
+                                                            Predicate<ConsumerRecord<K, V>> filter,
+                                                            long timeoutMs) {
+        if(offsets.isEmpty()) {
+            return RecordFetchResult.emptyResult();
+        }
+
+        Map<TopicPartition, OffsetRange> offsetRanges = fetchOffsetRanges(offsets.keySet());
+
+        offsets = filterOutUselessOffsetsReverseFetch(offsets, offsetRanges);
+        if (offsets.isEmpty()) {
+            return RecordFetchResult.emptyResult();
+        }
+
+        checkAndCorrectBounds(offsets, offsetRanges);
+
+        Map<TopicPartition, BiDirectionalRecordCollector> collectors =
+                initBiDirectionalRecordCollectorsForPartitions(offsets.keySet(), filter, limit);
+
+        assignConsumerToPartitions(consumer, offsets);
+
+        long fetchStart = System.currentTimeMillis();
+
+        Map<TopicPartition, OffsetRange> currentChunkOffsetRanges = 
+                getFirstChunkRanges(offsets, offsetRanges, collectors);
+
+        Map<TopicPartition, Long> consumedOffsets = new HashMap<>();
+
+        while(true) {
+
+            seekOffsetsOfCurrentChunkBeginning(consumer, currentChunkOffsetRanges);
+            collectRecordsFromCurrentChunk(consumer, timeoutMs, collectors, fetchStart,
+                                           currentChunkOffsetRanges, consumedOffsets);
+
+            getNextChunkRange(currentChunkOffsetRanges, offsetRanges, collectors);
+
+            if( allCurrentChunksIsEmpty(currentChunkOffsetRanges) ||
```

Comment:
This check does not consider timeout exhausting, so that iteration may continue again and again if check returns false. 
Is there any possibility of infinite iteration?

Reply:
Not a problem. No fix needed